docs/bir-spec/src/test/java/org/ballerinalang/birspec/BIRTestUtils.java

```diff
@@ -153,6 +158,73 @@ private static void assertFunctions(BIRNode.BIRPackage expectedBIR, Bir.Module b
             // assert basic blocks
             assertBasicBlocks(actualFunctionBody.functionBasicBlocksInfo(), expectedFunction.basicBlocks,
                     constantPoolEntries);
+
+            assertScopes(actualFunction.scopeEntries(), expectedFunction);
+        }
+    }
+
+    private static void assertScopes(ArrayList<Bir.ScopeEntry> scopeEntries, BIRNode.BIRFunction function) {
+        int instructionOffset = 0;
+        Map<Integer, ExpectedScopeEntry> scopes = new HashMap<>();
+        Set<BirScope> visitedScopes = new HashSet<>();
+
+        // Collect scope vs starting instruction offset
+        Collection<List<BIRNode.BIRBasicBlock>> basicBlocksCollection = function.parameters.values();
+        for (List<BIRNode.BIRBasicBlock> basicBlocks : basicBlocksCollection) {
+            instructionOffset = generateExpectedScopeEntries(basicBlocks, instructionOffset, scopes, visitedScopes);
+        }
+
+        generateExpectedScopeEntries(function.basicBlocks, instructionOffset, scopes, visitedScopes);
+
+        for (Bir.ScopeEntry actualScopeEntry : scopeEntries) {
+            ExpectedScopeEntry expectedScopeEntry = scopes.get(actualScopeEntry.currentScopeIndex());
+            Assert.assertNotNull(expectedScopeEntry);
+
+            Assert.assertEquals(actualScopeEntry.instructionOffset(), expectedScopeEntry.instructionOffset);
+            Assert.assertEquals(actualScopeEntry.hasParent(), expectedScopeEntry.hasParent);
+            Assert.assertEquals(actualScopeEntry.parentScopeIndex(), expectedScopeEntry.parentId);
+        }
+    }
+
+    private static int generateExpectedScopeEntries(List<BIRNode.BIRBasicBlock> bbList, int instructionOffset,
+            Map<Integer, ExpectedScopeEntry> scopes, Set<BirScope> visitedScopes) {
+        for (BIRNode.BIRBasicBlock bb : bbList) {
+            for (BIRAbstractInstruction instruction : bb.instructions) {
+                instructionOffset++;
+                BirScope instructionScope = instruction.scope;
+
+                if (!visitedScopes.contains(instructionScope)) {
```

Comment:
shall we inverse the if condition here? 

if (visitedScopes.contains(instructionScope)) {
    continue;
}

Reply:
Fixed